### PR TITLE
[pycbc live] Allowing the use of psd variation in the ranking statistic for pycbc live

### PR DESCRIPTION
**Pull request by ArthurTolley**
Changes made to `pycbc_live`:
- new argument for enabling psd_variation
- calling the `variation.py` module to perform the psd_var calculation
- creating filters
- saving psd_var values to results

changes made to `run.sh`:
- adding the psd_var argument and the new ranking statistic

changes made to `matchedfilter.py`:
- adding a psd_var values array to store the values calculated

changes made to `variation.py`:
- adding new functions for pycbc live specific psd_variation
- create_filter
- - creates a filter for each ifo using the current best psd
- calc_psd_variation
- - performs the convolution of the filter and the data to find the psd variation timeseries
- - average over 0.25s strides within this psdvar timeseries to remove the effect of any short duration glitches
- - remove any outliers from this 0.25s stride array
- - average further to retrieve a psd_var value for every second
- this follows the method used in the offline psd_var calculation
- find_var_value
- - take in the trigger end time and interpolate between the 2 psd_var values found for the seconds either side of the trigger. This is the final psd_var value for the statistic

### CODE REVIEW

**Comment by ArthurTolley** [comment]
Currently set to a draft PR:
- Have code comments to address
- Rename functions to be more explicitly pycbc live related
- Testing extensively all the changes

File: bin/pycbc_live
Comment (ArthurTolley) [inline bin/pycbc_live:1000]: Another way of 'turning on' the psd variation is to use the string provided to single ranking. 

I chose to do it this way to let psd var values be calculated and not used in the single ranking.

I think there does need to be a check within the parser args however to make sure that --psd-variation **is** given when somebody uses a single ranking that expects psd variation. Maybe this can be automatic?

File: bin/pycbc_live
Comment (ArthurTolley) [inline bin/pycbc_live:1179]: I don't particularly like the multiple `if args.psd_variation` statements thrown around at different points in this code but I can't think of how to generate these objects at the right time without them.

These could be generated regardless of the status of psd_variation but that feels wrong too.

File: bin/pycbc_live
Comment (ArthurTolley) [inline bin/pycbc_live:1261]: Not sure how many logging statements are needed. Thinking maybe just the single statement to say that PSD variation has been calculated for an ifo?

**Comment by tdent** [comment]
Not sure I can review all the aspects concerning psd var calculation and pycbc live internal workings .. can we have more reviewers for those?

**Comment by tdent** [comment]
Do we know why the 'run small search' test fails?

**Comment by tdent** [comment]
Codeclimate seems to be finding some nontrivial problems (undefined variable, unused variable) as well as some easily fixable (line too long) and ones that might require some thought (code duplication within psd variation module)

**Comment by titodalcanton** [comment]
Yup I will get to @ArthurTolley's PRs in the next days.

File: bin/pycbc_live
Comment (titodalcanton) [inline bin/pycbc_live:593]: Undo this change please.

File: pycbc/psd/variation.py
Comment (titodalcanton) [inline pycbc/psd/variation.py:221]: These two lines can be removed.

File: bin/pycbc_live
Comment (ArthurTolley) [reply bin/pycbc_live]: I've removed the first 2 `if args.psd_variation:` statements and had the `psd_recalculated` objects always be created and set whenever the psd has been recalculated.

**Comment by ArthurTolley** [comment]
@tdent I assigned you because I believe you worked on the original PSD variation for offline with Simone. I will be very thankful for review from Tito too.

Just made some changes to fix a lot of the codeclimate and to re-run the small search test, hopefully it will pass this time.

The code duplication in `variation.py` might just have to be a thing, the codes are similar enough to share specific functionality but different enough they require new functions for Live. This could be refactored but I didn't want to touch anything offline search related.

File: bin/pycbc_live
Comment (tdent) [reply bin/pycbc_live:1000]: mmm, so in order to _use_ the psd variation we need to also change the single stat, that is what you are saying?  That makes sense .. 
I suppose if you do that but don't calculate psd var you get an error, have you checked if the error is comprehensible?

File: bin/pycbc_live
Comment (tdent) [inline bin/pycbc_live:1178]: 'objects' -> 'dict'

What does this mean exactly, is it that if the range doesn't change by much the psd doesn't get recalculated at each new time segment?

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:302]: This doesn't quite make sense.  'This mean square' ?  And does taking the mean square by itself remove effects of short duration glitches ?  Reading the doc for the existing mean square function seems to say it is used to identify and remove outliers due to glitches.

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:362]: The comment is superfluous 
```suggestion
```

File: bin/pycbc_live
Comment (ArthurTolley) [reply bin/pycbc_live:1178]: Exactly, the arg `--psd-recalculate-difference` is used to determine if a new psd is too different from the current psd and the old one needs to be replaced ([link](https://github.com/gwastro/pycbc/blob/master/pycbc/strain/strain.py#L1681) to function).

We use the PSD to create the filter in the psd variation so need to keep track of when a new psd is being used by pycbc live. Will change it to say dict not just objects.

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:333]:
```suggestion
    # Grab the last increments worth of data, plus padding for edge effects.
```

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:302]: 'This mean square' -> 'The mean square'.

Yeah, the mean square identifies the outliers which are then removed and replaced. The mean square itself doesn't remove any effects of short duration glitches. I will reword this docstring.

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:337]: Typo 'Convolve'

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:334]: During testing I found that grabbing 3*data_trim toward the front of the strain was safer than only grabbing 1*data_trims worth.

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:304]: Slightly unclear wording .. well the number of values in the output is obvious if you read the docstring for 'increment' (it is a duration in seconds), so not necessary to explain here. I would just say 'to produce the PSD variation timeseries.'

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:311]: you can put a print type(full_filt) statement to find out what sort of object it is!

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:355]: this is a bit confusing - 'short_stride' is a time duration in seconds, so the variable you are calling 'stride' is an inverse time, or frequency.  This whole block probably needs a bit more explanation as to what all the different variables and indices are doing though (the original Simone code is not very clear either)

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:366]: Am gonna say this aspect of the line is also self-explanatory, so better without comment
```suggestion
```

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:311]: 🤦🏻‍♂️ I did this but forgot to update the docstring lol. Will do it now

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:348]:
```suggestion
    # Define an array of averages that is used to substitute outliers
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:367]: However, I am not sure why the 'epoch' given here is correct. Why does this look significantly different from the offline code?

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:378]:
```suggestion
    # Extract the PSD variation values at trigger times by linear interpolation
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:384]:
```suggestion
        Dictionary containing input trigger times.
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:399]:
```suggestion
    # Create the interpolator
```

File: bin/pycbc_live
Comment (spxiwh) [reply bin/pycbc_live:1000]: "so in order to use the psd variation we need to also change the single stat" -> Yes, that's Arthur's alternative proposal. I prefer this way. One might want to compute the PSD variation, but not actually use it. No other statistic is controlled by the single ranking (e.g. chi-squared isn't turned off if we choose ranking=SNR).

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:404]: Actually you don't need to create the variable 'trigger_times' as you only use it once. Can just do this directly IMO. (Then as the line is slightly more complex you can put a comment!)
```suggestion
    # Evaluate at the trigger times
    psd_var_vals = interpolator(triggers['end_time'])
```

**Review by tdent — CHANGES_REQUESTED** [review]
See comments on the variation.py code.  I am not so aware of the issues in pycbc_live with recalculating (or not) PSDs so I leave that to Tito or Alex. 

As the changes in matchedfilter.py are only blank lines and this is critical code (ie every change to it will have to be accounted for and reviewed) please revert them

File: bin/pycbc_live
Comment (tdent) [reply bin/pycbc_live:1000]: Sorry, I don't understand the difference here. It seems you are both saying the same thing

File: examples/live/run.sh
Comment (tdent) [inline examples/live/run.sh:208]: Verbosity comes last ;)
```suggestion
--psd-variation \
--verbose
```

File: pycbc/filter/matchedfilter.py
Comment (tdent) [inline pycbc/filter/matchedfilter.py:1665]:
```suggestion
```

File: bin/pycbc_live
Comment (spxiwh) [inline bin/pycbc_live:1264]: Similarly here, how can this ever not be None, 6 lines earlier it was reset to None, regardless of what happens further down at line 1282. Should `filt = {ifo: None for ifo in ifos}` also move outside of the main `while` loop? If so, it needs a more informative variable name.

File: pycbc/psd/variation.py
Comment (tdent) [reply pycbc/psd/variation.py:334]: Naively I would think 2* was enough ...

File: examples/live/run.sh
Comment (spxiwh) [inline examples/live/run.sh:208]: Please leave verbose as the last option, as often you want to turn it off in examples.

File: bin/pycbc_live
Comment (ArthurTolley) [reply bin/pycbc_live:1000]: > I suppose if you do that but don't calculate psd var you get an error, have you checked if the error is comprehensible?

From memory you would know you are missing psd_var values when calculating the statistic but you wouldn't know until a trigger came in. Which is why I would like to do an argument check to prevent running --sngl-stat '...psdvar' without having '--psd_variation' enabled.

File: pycbc/filter/matchedfilter.py
Comment (ArthurTolley) [reply pycbc/filter/matchedfilter.py:1674]: My mistake, I just deleted my changes rather than undoing them.

File: pycbc/psd/variation.py
Comment (spxiwh) [inline pycbc/psd/variation.py:134]: Why would this need `refcheck=False`? The array is only created in the line above, so it shouldn't be being referenced from anywhere else.

**Comment by ArthurTolley** [comment]
Sorry, had to reverse my latest commit due to including some incorrect `run.sh` changes. Will re-add all the other changes and re-commit to this branch.

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py]: I will change this because due to the data trimming this isn't even correct, it can be seen that it has increment + 2\*data_trim's worth of seconds.

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py]: I have updated the variable 'stride' to be 'samples_per_second' to help indicated what the variable actually represents. We want to calculate how many elements in 'short_ms' form 1 second's worth of data and then take an average of those samples.

I've also updated the comment to help:
"    # Calculate the PSD variation every second by averaging over short_ms every
    #  samples_per_second."

**Comment by ahnitz** [comment]
@ArthurTolley Out of curiosity, why does one not simply use the range already calculated for the PSD as function of time for the variation statistic, rather than do additional calculations / add more complexity? Is it because you want to calculate for very short time variations?

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:361]: From looking at the offline code, it is generating PSD var values for the whole 512 seconds of data and so it is logical to start from strain.start_time when defining the new timeseries.

In the live code we only want to find PSD var values for the newest increment of strain data so taking the end time of the data and subtracting the increment and the extra data from trimming is how I found the epoch of the psd var timeseries.

Example:
start_time = 0
end_time = 512
increment = 8
data_trim = 2

We take the latest data from 504 - 512, pad the start by 6 seconds and trim 2s from both sides. We calculate the mean square and average each second so we have 10 data points to put into our timeseries and we need to line the epoch up which is: end_time - increment - 2\*data_trim.

**Comment by ahnitz** [comment]
@ArthurTolley @tdent In terms of code duplication, it is worth thinking about how to remove that. I don't think we want to always postpone technical debt. What's the conceptual difference why the interface needs to be different?

File: bin/pycbc_live
Comment (ArthurTolley) [reply bin/pycbc_live:1264]: I have moved filt (and renamed) to outside of the main loop. I don't want this being reset at all once the search has commenced, only remade as the psd gets recalculated. I've also changed the logic on the remaking as I think it was wrong. Just testing now

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:134]: I don't remember exactly why I made this change but it was most likely due to it not working without the change. This could be very old. I will remove the refcheck in both the offline and live versions and see if there are any errors.

**Comment by spxiwh** [comment]
> @ArthurTolley @tdent In terms of code duplication, it is worth thinking about how to remove that. I don't think we want to always postpone technical debt. What's the conceptual difference why the interface needs to be different?

I am also worried about duplication here. I would prefer to merge this first, and have a working implementation of PSD variation in Live, and then have some combination of me and @GarethCabournDavies work through this module. It's not just about duplication, there are some parts of the original implementation that need to be more in line with the PyCBC interface. So I think we have a larger task here of "unify and standardize PSD variation code", but would prefer not to put that on a PhD student approaching completion date!

File: bin/pycbc_live
Comment (tdent) [reply bin/pycbc_live:1000]: Well you can do that, no?

**Comment by tdent** [comment]
@spxiwh Well I think the 7-line-plus-long-comment block which creates 'full_filt' is identical between the two routes .. and the comment is also identical .. so it would not take much work to extract this to an independent function, say `def make_weighted_filter` ?

Would not like to see the reasoning 'OK to duplicate code because xxx even if if refactoring would be easy' unless the reason is a _really_ tight deadline

File: bin/pycbc_live
Comment (ArthurTolley) [reply bin/pycbc_live:1000]: Yeah, it'll be in the arg verification step. I just need to figure out how to do it 😁 will be a job for tomorrow

**Comment by titodalcanton** [comment]
Still too distracted to properly sit and go through this, but I just wanted to make sure that the computational cost is not significantly affected by this change, at least with the example we have (a proper demonstration will need to be done on the MDC).

**Comment by ArthurTolley** [comment]
@titodalcanton I have been testing on MDC with a reduced template bank and have produced profiling plots. Would you like me to produce those and post them here?

**Comment by titodalcanton** [comment]
Profiling plots are good, but before going to those, I was asking about what lag (if you have been testing in actual low latency) or duty factor (if you have been running in high-latency or offline mode) the search reports after this change.

**Comment by ArthurTolley** [comment]
These tests were performed on the pycbc_live example in /examples/live using a template bank of 3,000 templates and 4 nodes. These are both done using 512s of live O3 Replay MDC data.

Without PSD_Variation:
Average Lag: 8.32842 s
Average Duty Factor: 0.954346

With PSD_Variation:
Average Lag: 7.34102 s
Average Duty Factor: 0.966953

I simply took the output and averaged all lag and duty factors. It can be seen that there is a negligible difference in the average lag and duty factor when psd_variation is enabled. Would you like to see more data or is this enough? Profiling plots showed very similar things, effectively no slowdown caused by psd_variation.

**Comment by titodalcanton** [comment]
Thanks @ArthurTolley, and how many detectors were producing triggers during those 512 s? If there all three were observing, then this is probably a good enough test for now.

**Comment by ArthurTolley** [comment]
@titodalcanton Should be 3 online according to [this lag plot](https://ldas-jobs.ligo.caltech.edu/~pycbc.live/pre_O4_mdc/full_bandwidth/2023/10/25/2023-10-25_lag_over_time.png) but, I think I might've made a mistake with my previous test. Sorry, let me try it again and report back the results

**Comment by ArthurTolley** [comment]
@titodalcanton sorry to spam you, the test **was** correct, I was looking at an old output file when I thought it was wrong. So yes, the lag and duty factors are correct for 3 detectors in live.

File: pycbc/psd/variation.py
Comment (tdent) [reply pycbc/psd/variation.py:361]: OK, so the calculation is done starting with data from 498 through 512, 2s are lost for filter artefacts giving 500 through 510, then the epoch is 1 'trim segment' later than the start of the original data, i.e. 2 rather than 3 'trims' plus the analysis stride ('increment').

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:344]: leftover line of comment? 
```suggestion
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:360]:
```suggestion

    m_s = numpy.array(m_s, dtype=wstrain.dtype)
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:379]: Where the timeseries came from is not relevant to understanding the function.
```suggestion
        latest increment in PyCBC Live.
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:385]:
```suggestion
        Array of interpolated PSD variation values at trigger times.
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:350]: I'm not sure if 'every samples_per_second' makes sense .. what do you want to say here?

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:134]: No errors from what I can see so I have removed it.

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:349]: Probably doesn't make sense once you convert 'sample_per_second' to a number. 

I want to say that we're averaging the short_ms array every X number of elements to find the average psd variation value each second. samples_per_second depends on the length of short_stride, if short_stride is 0.1 then it is every 10 samples/elements. If short_stride is 0.25 then it's 4 samples_per_second.

File: pycbc/psd/variation.py
Comment (ArthurTolley) [reply pycbc/psd/variation.py:358]: The epoch is 1 trim segment later than the start of the slice of the data we are using to calculate PSD variation, not the start of the original data because that varies depending on how much data PyCBC Live has in the buffer right now.

I could change the epoch to be `epoch=astrain.start_time + data_trim` if you would prefer? The change mainly being from strain -> astrain where astrain is the sliced strain data.

File: pycbc/events/coinc.py
Comment (tdent) [inline pycbc/events/coinc.py:998]:
```suggestion
            parser.error(f"The single ifo ranking stat {args.sngl_ranking}
                           requires --psd-variation.")
```

File: pycbc/psd/variation.py
Comment (tdent) [inline pycbc/psd/variation.py:241]:
```suggestion
        The current PyCBC Live PSD: variations are measured relative to this
        estimate.
```

File: pycbc/psd/variation.py
Comment (tdent) [reply pycbc/psd/variation.py:349]: oh, I think I see. See suggestion ..

File: pycbc/psd/variation.py
Comment (tdent) [reply pycbc/psd/variation.py:358]: I resolved the conversation as I didn't think there was a problem in the end.  But maybe it would be good to add a short comment to explain the time accounting, eg `The original data segment started at 'increment + 3 * data_trim' before the strain end_time, the variation time series starts 1 * data_trim later`

**Comment by tdent** [comment]
@ArthurTolley If you have 15 minutes or so to do the most obvious refactoring, could you make a function out of the the 7-line-plus-long-comment block which creates 'full_filt', which is identical between online and offline psd var (even the comment is the same) ?

**Comment by ArthurTolley** [comment]
@tdent I've moved filter creation to its own function, just going to test it now

**Comment by ArthurTolley** [comment]
It's working fine on my local test, will wait for PR tests to finish

**Comment by ArthurTolley** [comment]
@tdent In the future could you tag the lines you want me to change? That way I can respond to the change you're requesting and there won't be any confusion or time-wasting. I dislike us simply tagging each other in separate comments and making this 90+ comment PR even longer. I will do the changes you have requested now.

**Comment by tdent** [comment]
TBH I thought it would be unambiguous, eg if you search existing code for `full_filt`.

**Review by tdent — APPROVED** [review]
Looks OK to me, if tests etc. pass

**Comment by tdent** [comment]
Think this is ready to go, the remaining 'duplication' complaints are minor

**Comment by ArthurTolley** [comment]
Thank you @tdent, I'm glad you figured out that code climate line continuation, it was doing my head in.

@titodalcanton Would you like one final look or are you okay with this being merged?

**Comment by ArthurTolley** [comment]
I will merge this PR at the end of the day if no further comments are made 😁

**Pull request merged**
